docs/vfio-user.rst

```diff
@@ -204,12 +204,32 @@ A server can serve:
 1) one or more clients, and/or
 2) one or more virtual devices, belonging to one or more clients.
 
-The current protocol specification requires a dedicated socket per
-client/server connection. It is a server-side implementation detail whether a
-single server handles multiple virtual devices from the same or multiple
-clients. The location of the socket is implementation-specific. Multiplexing
-clients, devices, and servers over the same socket is not supported in this
-version of the protocol.
+The current protocol specification requires dedicated sockets per
+client/server connection. Commands in the client-to-server direction are
+handled on the main communication socket which the client connects to, and
+replies to these commands are passed on the same socket. Commands sent in the
+other direction from the server to the client as well as their corresponding
+replies, are exchanged on a separate socket, which is set up during negotiation
+(AF_UNIX servers just pass the file descriptor).
```

Comment:
IMO, we should settle on a single way of handling server->client commands. We have evidence that the current way is hard to implement correctly, and that the reference implementation is currently broken. I had assumed that the separate socket would become the only supported way for server->client commands. What's your take on this?

Separately, in most deployments we simply don't send any server->client commands. So, for these scenarios, it is strictly speaking still optional. But maybe that's not the sense of optional you were referring to?

Reply:
@jlevon WDYT?

Reply:
> I had assumed that the separate socket would become the only supported way for server->client commands. What's your take on this?

Requiring this means that we preclude future implementations that correctly handle this, so my vote would be to make this optional.

Reply:
that, plus for those (rather common) cases that actually don't send server->client, we shouldn't require an implementation on either side.

Reply:
> > I had assumed that the separate socket would become the only supported way for server->client commands. What's your take on this?
> 
> Requiring this means that we preclude future implementations that correctly handle this, so my vote would be to make this optional.

[I'll be using `twin_socket` as the feature name from now on per the other discussion with @jlevon ]

But then we're including two different ways to do the same thing in the specification. As an implementer, I will want to be interoperable with any peer, and that would force me to:
1. implement support for `twin_socket` and use it when available (since I can't assume the peer handles asynchronous messages correctly, given that not even the reference implementation does)
2. ensure my own implementation is completely async-safe (in case the peer doesn't implement `twin_socket`)

That seems a pretty unfortunate outcome. Perhaps we should mandate `twin_socket` support in server implementations, so clients can choose to specify `twin_socket` and simplify their socket I/O code as opposed to going full-async?

Another consequence of making `twin_socket` optional is that we would have to do something to fix the libvfio-user server implementation. At least it would be unsatisfactory to me to leave it known-broken for clients that don't use `twin_socket` and don't provide `mmap`-able DMA. Per the discussion on the issue, this would probably mean buffering of commands until we find the reply we're waiting for, with unfortunate consequence as discussed (no guaranteed limit for message buffers, socket back-pressure no longer works for flow control).

Reply:
(crossing messages)

I want to clarify that in the end this is your decision, and I'll be happy enough to get this working somehow, so take my comments as points to consider, and I will go with whatever decision you make :)

Reply:
It'd be totally fine for a client to refuse to work with a server that won't do twin_socket. This is a deployment-specific choice, not something that we should mandate *in the protocol*. For example, you certainly care, whereas for our use cases, we don't (hence it being broken all this time).

Re the non-twin-socket broken-ness in libvfio-user specifically: again, only some people care about this. I don't think it's necessary to try to fix it in non-twin-socket mode. What *would* make sense is some docs identifying the issue in this setup so people don't get caught out.

Reply:
OK, fair enough.

It probably makes sense then to have the default implementation print a warning to stderr in case it hits a code path that sends a command from server to client when `twin_socket` is off? Or even fail whatever operation is triggering that command?

Reply:
a stderr warning sounds good